### PR TITLE
Close underlying WebSocket connection when connection is closed

### LINKED ISSUE #7199: WebSocket protocol don't close the underlying IO
When called with a network address, `HTTP::WebSocket.new` delegates to `Protocol.new` which creates a `TCPSocket` and connects it to the given address. This socket is stored as an ivar in the `Protocol` instance and is never accessible from outside.

`WebSocket#close` sets a `closed` flag and sends a `close` message, but the underlying IO (`TCPSocket`) is never closed.

This can leak open TCP connections.

On the server-side implementation, `WebSocketHandler` explicitly calls `io.close` after the `WebSocket` session has finished. This call is not executed however, when the session handling raises.

See also #7197

### PR DESCRIPTION
**Pull request by willhbr**
This fixes #7199 by making `WebSocket::Protocol#close` call `TCPSocket#close`. `WebSocket#close` is called when an exception is raised, or when a CLOSE opcode is received.

Also wrapped `ws_session.run` in a begin...ensure to make sure that the `io` is closed on exceptions.

### CODE REVIEW

File: src/http/server/handlers/websocket_handler.cr [hunk 1]
```diff
@@ -40,8 +40,11 @@ class HTTP::WebSocketHandler
       response.upgrade do |io|
         ws_session = WebSocket.new(io)
         @proc.call(ws_session, context)
-        ws_session.run
-        io.close
+        begin
```
Comment (Sija) [inline src/http/server/handlers/websocket_handler.cr]: `begin ... end` is redundant here, `ensure` block would be just enough.

File: src/http/web_socket/protocol.cr [hunk 2]
```diff
@@ -232,10 +232,13 @@ class HTTP::WebSocket::Protocol
   end
 
   def close(message = nil)
-    if message
-      send(message.to_slice, Opcode::CLOSE)
-    else
-      send(Bytes.empty, Opcode::CLOSE)
+    unless @io.closed?
```
Comment (Sija) [inline src/http/web_socket/protocol.cr]: Prefer early-returns:

```suggestion
    return if @io.closed?
```

**Comment by RX14** [comment]
I think there should be a `sync_close` parameter (see `IO::Stapled` for an example) to make this behaviour configurable.

**Comment by willhbr** [comment]
Thanks for the review, I’m still working on this - just slowly

**Comment by willhbr** [comment]
Sadly I don't think I've got the time to work on this, so I'm going to close this for now. Will re-open if I have time in the future.

**Pull request closed without merging**